File: fetch_npm_downloads.py
#!/usr/bin/env python3

"""
Fetch NPM package download statistics for a specific date range
API: https://api.npmjs.org/downloads/point/{period}/{package}
"""

import sys
import json
import urllib.request
from datetime import datetime

def fetch_downloads(package_name, start_date, end_date):
    """Fetch download statistics from npm registry"""
    period = f"{start_date}:{end_date}"
    url = f"https://api.npmjs.org/downloads/point/{period}/{package_name}"

    print(f"Fetching downloads for {package_name} from {start_date} to {end_date}...")
    print(f"URL: {url}\n")

    try:
        with urllib.request.urlopen(url) as response:
            data = json.loads(response.read().decode())

            if 'downloads' in data:
                print("✓ Success!")
                print(f"Package: {data['package']}")
                print(f"Period: {data['start']} to {data['end']}")
                print(f"Total Downloads: {data['downloads']:,}")
                return data
            else:
                print("Response:", data)
                return None

    except urllib.error.HTTPError as e:
        print(f"HTTP Error {e.code}: {e.reason}")
        return None
    except Exception as e:
        print(f"Error: {str(e)}")
        return None

def validate_date(date_string):
    """Validate date format YYYY-MM-DD"""
    try:
        datetime.strptime(date_string, '%Y-%m-%d')
        return True
    except ValueError:
        return False

def main():
    if len(sys.argv) < 4:
        print("Usage: python fetch_npm_downloads.py <package-name> <start-date> <end-date>")
        print("Date format: YYYY-MM-DD")
        print("\nExample:")
        print("  python fetch_npm_downloads.py mcp-server-kubernetes 2025-11-27 2025-12-03")
        sys.exit(1)

    package_name = sys.argv[1]
    start_date = sys.argv[2]
    end_date = sys.argv[3]

    # Validate dates
    if not validate_date(start_date) or not validate_date(end_date):
        print("Error: Dates must be in YYYY-MM-DD format")
        sys.exit(1)

    fetch_downloads(package_name, start_date, end_date)

if __name__ == "__main__":
    main()

File: fetch_npm_downloads.js
#!/usr/bin/env node

/**
 * Fetch NPM package download statistics for a specific date range
 * API: https://api.npmjs.org/downloads/point/{period}/{package}
 */

const https = require('https');

function fetchDownloads(packageName, startDate, endDate) {
  const period = `${startDate}:${endDate}`;
  const url = `https://api.npmjs.org/downloads/point/${period}/${packageName}`;

  console.log(`Fetching downloads for ${packageName} from ${startDate} to ${endDate}...`);
  console.log(`URL: ${url}\n`);

  https.get(url, (res) => {
    let data = '';

    res.on('data', (chunk) => {
      data += chunk;
    });

    res.on('end', () => {
      try {
        const result = JSON.parse(data);

        if (result.downloads !== undefined) {
          console.log('✓ Success!');
          console.log(`Package: ${result.package}`);
          console.log(`Period: ${result.start} to ${result.end}`);
          console.log(`Total Downloads: ${result.downloads.toLocaleString()}`);
        } else {
          console.log('Response:', result);
        }
      } catch (error) {
        console.error('Error parsing response:', error.message);
        console.log('Raw response:', data);
      }
    });
  }).on('error', (error) => {
    console.error('Error fetching data:', error.message);
  });
}

// Parse command line arguments
const args = process.argv.slice(2);

if (args.length < 3) {
  console.log('Usage: node fetch_npm_downloads.js <package-name> <start-date> <end-date>');
  console.log('Date format: YYYY-MM-DD');
  console.log('\nExample:');
  console.log('  node fetch_npm_downloads.js mcp-server-kubernetes 2025-11-27 2025-12-03');
  process.exit(1);
}

const [packageName, startDate, endDate] = args;
fetchDownloads(packageName, startDate, endDate);
